### ISSUE TITLE
Inconsistent AST about `declare` keyword

### ISSUE THREAD

**Issue by mysticatea** [JamesHenry/typescript-estree#28]
**What version of TypeScript are you using?**

- 3.1.6

**What version of `typescript-estree` are you using?**

- 2.1.0 (typescript-eslint-parser#master)

**What code were you trying to parse?**

```ts
declare function f1(a: number): number
function f2(a: number): number
declare namespace n1 {}
namespace n2 {}
declare class C1 {}
class C2 {}

declare namespace N3 {
    declare function f1(a: number): number
    function f2(a: number): number
}
```

**What did you expect to happen?**

I can get consistent AST for semantics about whether the `declare` keyword exists or not.
For example, consistently `declare:boolean` property or consistently different AST type.

**What actually happened?**

<details>
<summary>I got:</summary>

```json
{
  "type": "Program",
  "body": [
    {
      "type": "TSEmptyBodyDeclareFunction",
      "id": {
        "type": "Identifier",
        "name": "f1"
      },
      "generator": false,
      "expression": false,
      "async": false,
      "params": [
        {
          "type": "Identifier",
          "name": "a",
          "typeAnnotation": {
            "type": "TSTypeAnnotation",
            "typeAnnotation": {
              "type": "TSNumberKeyword"
            }
          }
        }
      ],
      "body": null,
      "returnType": {
        "type": "TSTypeAnnotation",
        "typeAnnotation": {
          "type": "TSNumberKeyword"
        }
      }
    },
    {
      "type": "TSEmptyBodyFunctionDeclaration",
      "id": {
        "type": "Identifier",
        "name": "f2"
      },
      "generator": false,
      "expression": false,
      "async": false,
      "params": [
        {
          "type": "Identifier",
          "name": "a",
          "typeAnnotation": {
            "type": "TSTypeAnnotation",
            "typeAnnotation": {
              "type": "TSNumberKeyword"
            }
          }
        }
      ],
      "body": null,
      "returnType": {
        "type": "TSTypeAnnotation",
        "typeAnnotation": {
          "type": "TSNumberKeyword"
        }
      }
    },
    {
      "type": "TSModuleDeclaration",
      "id": {
        "type": "Identifier",
        "name": "n1"
      },
      "body": {
        "type": "TSModuleBlock",
        "body": []
      },
      "declare": true
    },
    {
      "type": "TSModuleDeclaration",
      "id": {
        "type": "Identifier",
        "name": "n2"
      },
      "body": {
        "type": "TSModuleBlock",
        "body": []
      }
    },
    {
      "type": "ClassDeclaration",
      "id": {
        "type": "Identifier",
        "name": "C1"
      },
      "body": {
        "type": "ClassBody",
        "body": []
      },
      "superClass": null
    },
    {
      "type": "ClassDeclaration",
      "id": {
        "type": "Identifier",
        "name": "C2"
      },
      "body": {
        "type": "ClassBody",
        "body": []
      },
      "superClass": null
    },
    {
      "type": "TSInterfaceDeclaration",
      "abstract": false,
      "body": {
        "type": "TSInterfaceBody",
        "body": []
      },
      "id": {
        "type": "Identifier",
        "name": "I1"
      },
      "heritage": []
    },
    {
      "type": "TSInterfaceDeclaration",
      "abstract": false,
      "body": {
        "type": "TSInterfaceBody",
        "body": []
      },
      "id": {
        "type": "Identifier",
        "name": "I2"
      },
      "heritage": []
    },
    {
      "type": "TSEnumDeclaration",
      "id": {
        "type": "Identifier",
        "name": "E1"
      },
      "members": [],
      "declare": true
    },
    {
      "type": "TSEnumDeclaration",
      "id": {
        "type": "Identifier",
        "name": "E2"
      },
      "members": []
    },
    {
      "type": "TSModuleDeclaration",
      "id": {
        "type": "Identifier",
        "name": "N3"
      },
      "body": {
        "type": "TSModuleBlock",
        "body": [
          {
            "type": "TSEmptyBodyDeclareFunction",
            "id": {
              "type": "Identifier",
              "name": "f1"
            },
            "generator": false,
            "expression": false,
            "async": false,
            "params": [
              {
                "type": "Identifier",
                "name": "a",
                "typeAnnotation": {
                  "type": "TSTypeAnnotation",
                  "typeAnnotation": {
                    "type": "TSNumberKeyword"
                  }
                }
              }
            ],
            "body": null,
            "returnType": {
              "type": "TSTypeAnnotation",
              "typeAnnotation": {
                "type": "TSNumberKeyword"
              }
            }
          },
          {
            "type": "TSEmptyBodyFunctionDeclaration",
            "id": {
              "type": "Identifier",
              "name": "f2"
            },
            "generator": false,
            "expression": false,
            "async": false,
            "params": [
              {
                "type": "Identifier",
                "name": "a",
                "typeAnnotation": {
                  "type": "TSTypeAnnotation",
                  "typeAnnotation": {
                    "type": "TSNumberKeyword"
                  }
                }
              }
            ],
            "body": null,
            "returnType": {
              "type": "TSTypeAnnotation",
              "typeAnnotation": {
                "type": "TSNumberKeyword"
              }
            }
          }
        ]
      },
      "declare": true
    }
  ],
  "sourceType": "script"
}
```

</details><br>

In short:

- On functions, the `declare` keyword changes AST node types: `TSEmptyBodyDeclareFunction` or `TSEmptyBodyFunctionDeclaration`
- On classes, the `declare` keyword does not change AST. I cannot distinguish it.
- On interfaces, the `declare` keyword does not change AST. I cannot distinguish it.
- On enums, the `declare` keyword changes `node.declare` property.
- On namespaces, the `declare` keyword changes `node.declare` property.
  - In namespaces which have the `declare` keyword, in semantics, I think that the declarations in the namespaces are same as having the `declare` keyword, but...

Anyway, I want a consistent behavior around this.

**Comment by j-f1** [comment]
See also https://github.com/eslint/typescript-eslint-parser/issues/384

**Comment by mysticatea** [comment]
`TSAmbientDeclaration` may be nice.

```ts
interface TSAmbientDeclaration extends Statement {
    declaration:
        | VariableDeclaration
        | FunctionDeclaration
        | TSInterfaceDeclaration
        | TSEnumDeclaration
        | TSModuleDeclaration
}
```

All declaration in this node, contains declarations in module blocks recursively, are handled as ambient declarations.

This is similar to `Export*Declaration` in ESTree, `TSAmbientDeclaration` will be exported from this file.

**Comment by armano2** [comment]
@JamesHenry can you take a look on this?

**Comment by JamesHenry** [comment]
I won’t have time over the next week as I’m travelling a lot. Definitely open to PRs if anyone else gets chance, otherwise I may have some time the week after.

My gut reaction is: I prefer @mystictea’s approach but that initially alignment with Babel might be preferable. We can then consider if it’s worth updating both tools.

WDYT?

**Comment by armano2** [comment]
i can make those changes, but i don't know which one to go for... :>

**Comment by JamesHenry** [comment]
Let’s go for matching Babel unless there is a strong reason not to as a general rule. It’s why I spent the time setting up the testing for it. The Babel support was built by the TS Team themselves and they have always been open to changes where they make sense, so if we feel strongly the AST design is not optimal we can propose a change to them and carry it out in both tools later

**Comment by mysticatea** [comment]
I follow the @JamesHenry decision because this is your product :smile: .

For what it's worth, as I said on https://github.com/JamesHenry/typescript-estree/pull/38#issuecomment-443058786, JS and TS have very different mental model about AST. I'm warried that the mix of the two in `typescript-estree` causes confusion.

**Comment by armano2** [comment]
@mysticatea i did most of changes for delcare, but i'm unsure what i should do with
```
type name = string
```

right now it`s
```
  ],
  "sourceType": "script",
  "type": "Program"
}
```

but babel defines is at new node `TSTypeAliasDeclaration`

**Comment by armano2** [comment]
There is also this `DeclareFunction` - function without body 
but babel defines is as `TSDeclareFunction`

**Comment by mysticatea** [comment]
I remember https://github.com/eslint/typescript-eslint-parser/issues/414

**Comment by armano2** [comment]
ok, i will take a look on this to than :>

**Comment by armano2** [comment]
@JamesHenry this issue got solved in #42
